File: scholarzim/src/main/java/com/scholarzim/controller/SavedScholarshipController.java
package com.scholarzim.controller;

import com.scholarzim.service.SavedScholarshipService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.lang.NonNull;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import java.util.Collections;


@Slf4j
@Controller
public class SavedScholarshipController {

    private final SavedScholarshipService savedScholarshipService;

    public SavedScholarshipController(SavedScholarshipService savedScholarshipService) {
        this.savedScholarshipService = savedScholarshipService;
    }

    @GetMapping("/applicant/saved")
    public String listSaved(@NonNull Authentication auth, Model model) {
        try {
            var saved = savedScholarshipService.listSaved(auth.getName());
            model.addAttribute("saved", saved);
            model.addAttribute("loadFailed", false);
            // #region agent log
            String name = auth.getName() != null ? auth.getName() : "";
            boolean demoApplicant = "tanaka.moyo@student.co.zw".equalsIgnoreCase(name)
                    || "rudo.chikomo@student.co.zw".equalsIgnoreCase(name)
                    || "simba.ndlovu@student.co.zw".equalsIgnoreCase(name);
            com.scholarzim.debug.AgentDebugLog.log("C", "SavedScholarshipController.listSaved", "saved_ok",
                    java.util.Map.of(
                            "count", saved.size(),
                            "empty", saved.isEmpty(),
                            "isDemoApplicant", demoApplicant));
            // #endregion
            // #region agent log
            if (saved.isEmpty()) {
                com.scholarzim.debug.AgentDebugLog.log("D", "SavedScholarshipController.listSaved",
                        "saved_empty_shows_empty_state",
                        java.util.Map.of("count", 0, "isDemoApplicant", demoApplicant));
            }
            // #endregion
        } catch (Exception ex) {
            log.warn("Saved scholarships list failed for {}: {}", auth.getName(), ex.getMessage());
            model.addAttribute("saved", Collections.emptyList());
            model.addAttribute("loadFailed", true);
            // #region agent log
            com.scholarzim.debug.AgentDebugLog.log("C", "SavedScholarshipController.listSaved", "saved_failed",
                    java.util.Map.of(
                            "exClass", ex.getClass().getName(),
                            "exMessage", String.valueOf(ex.getMessage())));
            // #endregion
        }
        return "applicant/saved";
    }

    @PostMapping("/applicant/saved/{id}")
    public String save(
            @PathVariable Long id,
            @NonNull Authentication auth,
            RedirectAttributes redirect) {

        try {
            savedScholarshipService.save(auth.getName(), id);
            redirect.addFlashAttribute("successMessage", "Scholarship saved.");
        } catch (Exception ex) {
            log.warn("Save scholarship failed for {}: {}", auth.getName(), ex.getMessage());
            redirect.addFlashAttribute("errorMessage", "Could not save scholarship. Please try again.");
        }
        return "redirect:/scholarships/" + id;
    }

    @PostMapping("/applicant/saved/{id}/remove")
    public String remove(
            @PathVariable Long id,
            @NonNull Authentication auth,
            RedirectAttributes redirect) {

        try {
            savedScholarshipService.remove(auth.getName(), id);
            redirect.addFlashAttribute("successMessage", "Removed from saved list.");
        } catch (Exception ex) {
            log.warn("Remove saved scholarship failed for {}: {}", auth.getName(), ex.getMessage());
            redirect.addFlashAttribute("errorMessage", "Could not remove saved scholarship.");
        }
        return "redirect:/applicant/saved";
    }
}

File: app/Http/Controllers/Applicant/SavedScholarshipController.php
<?php

namespace App\Http\Controllers\Applicant;

use App\Http\Controllers\Controller;
use App\Services\ApplicationService;
use App\Services\SavedScholarshipService;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;

class SavedScholarshipController extends Controller
{
    public function __construct(
        private readonly SavedScholarshipService $savedScholarshipService,
        private readonly ApplicationService $applicationService,
    ) {
    }

    public function index(Request $request)
    {
        return view('applicant.saved', [
            'saved' => $this->savedScholarshipService->listSaved($request->user()),
            'appliedIds' => $this->applicationService->appliedIds($request->user()),
            'awards' => $this->applicationService->awardsByOpportunity($request->user()),
        ]);
    }

    public function store(Request $request, int $id)
    {
        try {
            $this->savedScholarshipService->save($request->user(), $id);
            $message = ['successMessage', 'Scholarship saved.'];
        } catch (\Throwable $e) {
            Log::warning('Save scholarship failed', ['user' => $request->user()->email, 'error' => $e->getMessage()]);
            $message = ['errorMessage', 'Could not save scholarship. Please try again.'];
        }

        return back()->with(...$message);
    }

    public function destroy(Request $request, int $id)
    {
        try {
            $this->savedScholarshipService->remove($request->user(), $id);
            $message = ['successMessage', 'Removed from saved list.'];
        } catch (\Throwable $e) {
            Log::warning('Remove saved scholarship failed', ['user' => $request->user()->email, 'error' => $e->getMessage()]);
            $message = ['errorMessage', 'Could not remove saved scholarship.'];
        }

        return back()->with(...$message);
    }
}
